### PR TITLE
Gatsby v4 upgrade [Fixes #4271]

### LINKED ISSUE #4271: Upgrade Gatsby to v4
**Is your feature request related to a problem? Please describe.**

Let's upgrade Gatsby to v4! Should help us reduce build times 🤞 

**Describe the solution you'd like**

https://www.gatsbyjs.com/docs/reference/release-notes/migrating-from-v3-to-v4/

### PR DESCRIPTION
**Pull request by wackerow**
## Description
- [x] Upgrade codebase to use recently released Gatsby v4
- [x] Upgrade other project dependencies for compatibility (includes `gatsby-plugin-image`)
- [x] Update code base with new syntaxt for images to support `gatsby-plugin-image`
- [x] Test builds on forked Gatsby Cloud account -- Builds passing
- [x] Lambda function builds moved to `public/functions` for compatibility between Gatsby Cloud builds and Netlify serving
- [x] DevOps to set up EF account for Gatsby Cloud
- [x] Set up Gatsby Cloud to build all PR branches for `dev` branch (including `dev`)
- [x] Set up Gatsby Cloud to build our deploy PRs from `dev` into `master`, which will build final production build from `master` branch.
- [ ] Set up Gatby Cloud account to _host_ final production build of `master` branch using existing Netlify account.
- [ ] Update to serve this build as `ethereum.org`

## Related Issue
#4271

### CODE REVIEW

**Comment by pettinarip** [comment]
@wackerow I've changed the config for `gatsby-plugin-sitemap` to support the newer version. The problem is the following:

I've used the default configs from the plugin. This means that the `xml`s files are located by default under the `/sitemap` folder.

- Prev version: `https://ethereum.org/sitemap.xml`
- Current version: `https://ethereum.org/sitemap/sitemap-index.xml` and `https://ethereum.org/sitemap/sitemap-0.xml`

Things to consider:
- If we want to keep this new path, we need to make sure that the Google Search Console is reading this path.
- We could do this:
```js
options: {
    output: `/`, // the default value is `/sitemap`
    ...
}
```
- This way, we could have the sitemap like this: `https://ethereum.org/sitemap-index.xml` and `https://ethereum.org/sitemap-0.xml`

File: netlify.toml [hunk 1]
```diff
@@ -14,13 +14,14 @@
   command = "yarn build" # debug with: "yarn build --verbose"
 
   # Directory with the serverless Lambda functions to deploy to AWS.
-  functions = "public/lambda"
+  functions = "public/functions"
```
Comment (samajammin) [inline netlify.toml:17]: Do we want to change this? Would we want to rename the directory in the file system also? Or is this a Gatsby v4 thing?

Comment (samajammin) [reply netlify.toml:17]: All our functions are still in `/src/lambda/` currently.

File: gatsby-node.js [hunk 2]
```diff
@@ -76,7 +80,6 @@ const checkIsMdxOutdated = (path) => {
       intlMatch += match.replace(re, (_, p1, p2) => p1 + p2)
     })
   } catch {
-    console.error("regex error")
```
Comment (samajammin) [inline gatsby-node.js:79]: I believe we still want this error log.

File: src/content/developers/tutorials/uniswap-v2-annotated-code/index.md [hunk 3]
```diff
@@ -382,7 +382,7 @@ This function allows the factory (and only the factory) to specify the two ERC-2
 
 #### Internal Update Functions {#pair-update-internal}
 
-##### \_update {#_update}
+##### \_update {#\_update}
```
Comment (samajammin) [inline src/content/developers/tutorials/uniswap-v2-annotated-code/index.md:385]: Not sure why this change is here - what's the purpose?

File: src/content/developers/tutorials/uniswap-v2-annotated-code/index.md [hunk 4]
```diff
@@ -444,7 +444,7 @@ This price calculation is the reason we need to know the old reserve sizes.
 
 Finally, update the global variables and emit a `Sync` event.
 
-##### \_mintFee {#_mintfee}
+##### \_mintFee {#\_mintfee}
```
Comment (samajammin) [inline src/content/developers/tutorials/uniswap-v2-annotated-code/index.md:447]: Same here

File: netlify.toml (hunk 1 above)
Comment (pettinarip) [reply netlify.toml:17]: Yes, this is something we need to change due that we now deploy them from GC. More details here: https://support.gatsbyjs.com/hc/en-us/articles/360054529274-Deploying-Netlify-Functions-from-Gatsby-Cloud

Our source functions still live in `src/lambda`, what has changed is the destination of the compiled functions, `public/functions`. And that destination is the default destination used by Netlify.

Note: I tried keeping this untouched but Netlify wasn't able to find the functions when deployed from GC. It seems that this configuration only works in their CI/CD pipeline (when you link a GH repo and run the builds in their platform). 

Note 2: Besides that, I'm not sure even if the `netlify.toml` is being uploaded from GC. I have the feeling that only the content of the `/public` folder is transferred and Netlify only works as a plain hosting service.

File: src/content/developers/tutorials/uniswap-v2-annotated-code/index.md (hunk 3 above)
Comment (wackerow) [reply src/content/developers/tutorials/uniswap-v2-annotated-code/index.md:385]: Hm, not sure. This appears to have been pulled in when merging `dev` back in, however, noticing that our markdown H5's don't parse these custom headers appropriately as is:

image
image

I feel we should just remove these H5 custom header ids until we can address this issue separately. Can also remove the stray escape `\` that seems to be breaking the others, ie:
image

File: src/content/developers/tutorials/uniswap-v2-annotated-code/index.md (hunk 4 above)
Comment (wackerow) [reply src/content/developers/tutorials/uniswap-v2-annotated-code/index.md:447]: See above

File: gatsby-node.js (hunk 2 above)
Comment (wackerow) [reply gatsby-node.js:79]: Took this out after it was just producing a handful of non-specific errors in the build console. Will replace it with 
```console.error(`regex error in ${englishPath}`)```

Comment (wackerow) [reply gatsby-node.js:79]: After chatting with @pettinarip decided to change this to a `console.warn` instead
